lib/chef/telemetry/decision.rb

```diff
@@ -1,12 +1,91 @@
 require "chef-config/path_helper"
 require "chef-config/windows"
+require "logger"
+require "forwardable"
+
+require_relative "decision/argument"
+require_relative "decision/environment"
+require_relative "decision/file"
+require_relative "decision/prompt"
 
 # Decision allows us to inspect whether the user has made a decision to opt in or opt out of telemetry.
 class Chef
   class Telemetry
-    module Decision
-      OPT_OUT_FILE = "telemetry_opt_out".freeze
-      OPT_IN_FILE = "telemetry_opt_in".freeze
+    class Decision
+      extend Forwardable
+
+      attr_reader :config, :enabled, :arg_decision, :env_decision, :file_decision, :prompt_decision, :logger
+
+      def initialize(opts = {})
+        @config = opts
+        @logger = opts[:logger] || Logger.new(opts.key?(:output) ? opts[:output] : STDERR)
+        @config[:output] ||= STDOUT
+        config[:logger] = logger
+
+        # This is the whole point - whether telemetry shoud be enabled in this
+        # runtime invocation. Start by assuming no.
+        @enabled = false
+
+        # The various things that have a say in whether telemetry should be enabled.
+        @arg_decision = Decision::Argument.new(ARGV)
+        @env_decision = Decision::Environment.new(ENV)
+        @file_decision = Decision::File.new(config)
+        @prompt_decision = Decision::Prompt.new(config)
+      end
+
+      #
+      # Methods for obtaining consent from the user.
+      #
+      def check_and_persist(dir = Decision::File.default_file_location)
+        file_decision.local_dir = dir
+
+        # If a decision is made by CLI arg, set runtime decision and do not persist
+        logger.debug "Telemetry decision examining CLI arg checks"
+        return @enabled = true if @arg_decision.enable?
+        return @enabled = false if @arg_decision.disable?
+
+        # If a non-persisting decision is made by env, only set runtime decision
+        logger.debug "Telemetry decision examining ephemeral ENV checks"
+        return @enabled = true if @env_decision.opt_in_no_persist?
+        return @enabled = false if @env_decision.opt_out_no_persist?
+
+        # Check to see if a persistent decision is made by env but not yet persisted
+        #  then persist it and set runtime decision
+        logger.debug "Telemetry decision examining persistent ENV checks"
+        if env_decision.opt_in?
+          file_decision.persist(true, dir) if !persisted? || file_decision.opt_out?
```

Comment:
I want to make sure I am understanding this logic. If we set opt-in and persist with an environment variable, we persist that value if 1) it has not been persisted before or 2) if they previously had an opt-out decision recorded in the file.

So I'm understanding that the environment variable opt-in will overwrite the previous file-based opt-out, and it will persist that new value, correct?

Reply:
That is correct, for certain values of environment variables. The -no-persist variants do not persist.

All code within `check_and_persist` is totally open to revision per the product team, anyway - I expect changes in logic here so ... just looking for someone to say what's what.